[dynamo] Fix traceback generation on runtime errors

xmfan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122535
* #122353
* __->__ #122746
* #122691

Fixes `During handling of the above exception, another exception occurred: [...] torch._dynamo.exc.Unsupported: generator`. traceback.format_exc uses generators which isn't supported by dynamo yet. 
<details>
  <summary>current error message</summary>

```
======================================================================                                                                                                                                                                                                                                                                                
ERROR: test_custom_fn_saved_tensors (__main__.TestCompiledAutograd)                                                                                                        
----------------------------------------------------------------------                                                                                                                                                                                                                                                                                
Traceback (most recent call last):                                                                                                                                                                                                                                                                                                                    
  File "/home/xmfan/core/pytorch/torch/fx/graph_module.py", line 307, in __call__                                                                                                                                                                                                                                                                     
    return super(self.cls, obj).__call__(*args, **kwargs)  # type: ignore[misc]                                                                                            
  File "/home/xmfan/core/pytorch/torch/nn/modules/module.py", line 1527, in _wrapped_call_impl                                                                             
    return self._call_impl(*args, **kwargs)                                                                                                                                                                                                                                                                                                           
  File "/home/xmfan/core/pytorch/torch/nn/modules/module.py", line 1537, in _call_impl                                                                                                                                                                                                                                                                
    return forward_call(*args, **kwargs)                                                                                                                                                                                                                                                                                                              
  File "<eval_with_key>.0", line 4, in forward                                                                                                                                                                                                                                                                                                        
    def forward(self, inputs, sizes, hooks):                                                                                                                                                                                                                                                                                                          
IndexError: list index out of range                                                                                                                                                                                                                                                                                                                   
                                                                                                                                                                                                                                                                                                                                                      
During handling of the above exception, another exception occurred:                                                                                                                                                                                                                                                                                   
                                                                                                                                                                                                                                                                                                                                                      
Traceback (most recent call last):                                                                                                                                                                                                                                                                                                                    
  File "/home/xmfan/core/pytorch/torch/testing/_internal/common_utils.py", line 2741, in wrapper                                                                                                                                                                                                                                                      
    method(*args, **kwargs)                                                                                                                                                                                                                                                                                                                           
  File "/home/xmfan/core/pytorch/test/inductor/test_compiled_autograd.py", line 499, in test_custom_fn_saved_tensors                                                                                                                                                                                                                                  
    self.check_output_and_recompiles(fn, 1)                                                                                                                                                                                                                                                                                                           
  File "/home/xmfan/core/pytorch/test/inductor/test_compiled_autograd.py", line 61, in check_output_and_recompiles                                                                                                                                                                                                                                    
    actual = list(opt_fn())                                                                                                                                                                                                                                                                                                                           
  File "/home/xmfan/core/pytorch/test/inductor/test_compiled_autograd.py", line 495, in fn                                                                                                                                                                                                                                                            
    loss.backward()                                                                                                                                                                                                                                                                                                                                   
  File "/home/xmfan/core/pytorch/torch/_tensor.py", line 534, in backward                                                                                                                                                                                                                                                                             
    torch.autograd.backward(                                                                                                                                                                                                                                                                                                                          
  File "/home/xmfan/core/pytorch/torch/autograd/__init__.py", line 267, in backward                                                                                                                                                                                                                                                                   
    _engine_run_backward(                                                                                                                                                                                                                                                                                                                             
  File "/home/xmfan/core/pytorch/torch/autograd/graph.py", line 766, in _engine_run_backward                                                                                                                                                                                                                                                          
    return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass                                                                                                                                                                                                                                             
  File "/home/xmfan/core/pytorch/torch/nn/modules/module.py", line 1527, in _wrapped_call_impl                                                                             
    return self._call_impl(*args, **kwargs)                                          
  File "/home/xmfan/core/pytorch/torch/nn/modules/module.py", line 1537, in _call_impl                                                                                     
    return forward_call(*args, **kwargs)                                             
  File "/home/xmfan/core/pytorch/torch/_dynamo/eval_frame.py", line 397, in _fn                                                                                            
    res = fn(*args, **kwargs)                                                        
  File "/home/xmfan/core/pytorch/torch/fx/graph_module.py", line 741, in call_wrapped                                                                                      
    return self._wrapped_call(self, *args, **kwargs)                                 
  File "/home/xmfan/core/pytorch/torch/fx/graph_module.py", line 315, in __call__                                                                                          
    _WrappedCall._generate_error_message(topmost_framesummary),                      
  File "/home/xmfan/core/pytorch/torch/fx/graph_module.py", line 289, in _generate_error_message                                                                           
    tb_repr = get_traceback()                                                        
  File "/home/xmfan/core/pytorch/torch/fx/graph_module.py", line 288, in get_traceback                                                                                     
    return traceback.format_exc()                                                    
  File "/home/xmfan/.conda/envs/benchmarks/lib/python3.10/traceback.py", line 183, in format_exc                                                                           
    return "".join(format_exception(*sys.exc_info(), limit=limit, chain=chain))                                                                                            
  File "/home/xmfan/.conda/envs/benchmarks/lib/python3.10/traceback.py", line 136, in format_exception                                                                     
    return list(te.format(chain=chain))                                              
  File "/home/xmfan/core/pytorch/torch/_dynamo/convert_frame.py", line 941, in catch_errors                                                                                
    return callback(frame, cache_entry, hooks, frame_state, skip=1)                  
  File "/home/xmfan/core/pytorch/torch/_dynamo/convert_frame.py", line 348, in _convert_frame_assert                                                                       
    unimplemented("generator")                                                       
  File "/home/xmfan/core/pytorch/torch/_dynamo/exc.py", line 199, in unimplemented                                                                                         
    raise Unsupported(msg)                                                           
torch._dynamo.exc.Unsupported: generator
```

</details>


With this change, we get back the descriptive error message:
<details>
  <summary>post-fix error message</summary>

```
Traceback (most recent call last):                 
  File "/home/xmfan/core/pytorch/torch/fx/graph_module.py", line 307, in __call__
    return super(self.cls, obj).__call__(*args, **kwargs)  # type: ignore[misc]
  File "/home/xmfan/core/pytorch/torch/nn/modules/module.py", line 1527, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)                                                                                                  
  File "/home/xmfan/core/pytorch/torch/nn/modules/module.py", line 1537, in _call_impl
    return forward_call(*args, **kwargs)                                                                                                     
  File "<eval_with_key>.0", line 4, in forward
    def forward(self, inputs, sizes, hooks):                                                                                                 
IndexError: list index out of range
                                                                                                                                             
Call using an FX-traced Module, line 4 of the traced Module's generated forward function:
                                                                                                                                             
def forward(self, inputs, sizes, hooks):
                                                                                                                                             
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
    getitem = inputs[0]                                                                                                                      
                                                                      
    getitem_1 = inputs[1];  inputs = None      
```

</details>

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

xmfan committed:
[dynamo] fix traceback generation on runtime errors

[ghstack-poisoned]

xmfan commented:
@pytorchbot merge

anijain2305 commented on torch/fx/graph_module.py:
```diff
@@ -283,7 +283,10 @@ def _generate_error_message(frame_summary: traceback.FrameSummary) -> str:
+        @torch._dynamo.disable()
+        def get_traceback():
+            return traceback.format_exc()
+        tb_repr = get_traceback()
```
Nit - You can do `tb_repr = torch._dynamo.disable(traceback.format_exc)()`

anijain2305 reviewed (approved):
Very nice!

xmfan committed:
Update on "[dynamo] Fix traceback generation on runtime errors"


Fixes `During handling of the above exception, another exception occurred: [...] torch._dynamo.exc.Unsupported: generator`. traceback.format_exc uses generators which isn't supported by dynamo yet. 
<details>
  <summary>current error message</summary>

```
======================================================================                                                                                                                                                                                                                                                                                
ERROR: test_custom_fn_saved_tensors (__main__.TestCompiledAutograd)                                                                                                        
----------------------------------------------------------------------                                                                                                                                                                                                                                                                                
Traceback (most recent call last):                                                                                                                                                                                                                                                                                                                    
  File "/home/xmfan/core/pytorch/torch/fx/graph_module.py", line 307, in __call__                                                                                                                                                                                                                                                                     
    return super(self.cls, obj).__call__(*args, **kwargs)  # type: ignore[misc]                                                                                            
  File "/home/xmfan/core/pytorch/torch/nn/modules/module.py", line 1527, in _wrapped_call_impl                                                                             
    return self._call_impl(*args, **kwargs)                                                                                                                                                                                                                                                                                                           
  File "/home/xmfan/core/pytorch/torch/nn/modules/module.py", line 1537, in _call_impl                                                                                                                                                                                                                                                                
    return forward_call(*args, **kwargs)                                                                                                                                                                                                                                                                                                              
  File "<eval_with_key>.0", line 4, in forward                                                                                                                                                                                                                                                                                                        
    def forward(self, inputs, sizes, hooks):                                                                                                                                                                                                                                                                                                          
IndexError: list index out of range                                                                                                                                                                                                                                                                                                                   
                                                                                                                                                                                                                                                                                                                                                      
During handling of the above exception, another exception occurred:                                                                                                                                                                                                                                                                                   
                                                                                                                                                                                                                                                                                                                                                      
Traceback (most recent call last):                                                                                                                                                                                                                                                                                                                    
  File "/home/xmfan/core/pytorch/torch/testing/_internal/common_utils.py", line 2741, in wrapper                                                                                                                                                                                                                                                      
    method(*args, **kwargs)                                                                                                                                                                                                                                                                                                                           
  File "/home/xmfan/core/pytorch/test/inductor/test_compiled_autograd.py", line 499, in test_custom_fn_saved_tensors                                                                                                                                                                                                                                  
    self.check_output_and_recompiles(fn, 1)                                                                                                                                                                                                                                                                                                           
  File "/home/xmfan/core/pytorch/test/inductor/test_compiled_autograd.py", line 61, in check_output_and_recompiles                                                                                                                                                                                                                                    
    actual = list(opt_fn())                                                                                                                                                                                                                                                                                                                           
  File "/home/xmfan/core/pytorch/test/inductor/test_compiled_autograd.py", line 495, in fn                                                                                                                                                                                                                                                            
    loss.backward()                                                                                                                                                                                                                                                                                                                                   
  File "/home/xmfan/core/pytorch/torch/_tensor.py", line 534, in backward                                                                                                                                                                                                                                                                             
    torch.autograd.backward(                                                                                                                                                                                                                                                                                                                          
  File "/home/xmfan/core/pytorch/torch/autograd/__init__.py", line 267, in backward                                                                                                                                                                                                                                                                   
    _engine_run_backward(                                                                                                                                                                                                                                                                                                                             
  File "/home/xmfan/core/pytorch/torch/autograd/graph.py", line 766, in _engine_run_backward                                                                                                                                                                                                                                                          
    return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass                                                                                                                                                                                                                                             
  File "/home/xmfan/core/pytorch/torch/nn/modules/module.py", line 1527, in _wrapped_call_impl                                                                             
    return self._call_impl(*args, **kwargs)                                          
  File "/home/xmfan/core/pytorch/torch/nn/modules/module.py", line 1537, in _call_impl                                                                                     
    return forward_call(*args, **kwargs)                                             
  File "/home/xmfan/core/pytorch/torch/_dynamo/eval_frame.py", line 397, in _fn                                                                                            
    res = fn(*args, **kwargs)                                                        
  File "/home/xmfan/core/pytorch/torch/fx/graph_module.py", line 741, in call_wrapped                                                                                      
    return self._wrapped_call(self, *args, **kwargs)                                 
  File "/home/xmfan/core/pytorch/torch/fx/graph_module.py", line 315, in __call__                                                                                          
    _WrappedCall._generate_error_message(topmost_framesummary),                      
  File "/home/xmfan/core/pytorch/torch/fx/graph_module.py", line 289, in _generate_error_message                                                                           
    tb_repr = get_traceback()                                                        
  File "/home/xmfan/core/pytorch/torch/fx/graph_module.py", line 288, in get_traceback                                                                                     
    return traceback.format_exc()                                                    
  File "/home/xmfan/.conda/envs/benchmarks/lib/python3.10/traceback.py", line 183, in format_exc                                                                           
    return "".join(format_exception(*sys.exc_info(), limit=limit, chain=chain))                                                                                            
  File "/home/xmfan/.conda/envs/benchmarks/lib/python3.10/traceback.py", line 136, in format_exception                                                                     
    return list(te.format(chain=chain))                                              
  File "/home/xmfan/core/pytorch/torch/_dynamo/convert_frame.py", line 941, in catch_errors                                                                                
    return callback(frame, cache_entry, hooks, frame_state, skip=1)                  
  File "/home/xmfan/core/pytorch/torch/_dynamo/convert_frame.py", line 348, in _convert_frame_assert                                                                       
    unimplemented("generator")                                                       
  File "/home/xmfan/core/pytorch/torch/_dynamo/exc.py", line 199, in unimplemented                                                                                         
    raise Unsupported(msg)                                                           
torch._dynamo.exc.Unsupported: generator
```

</details>


With this change, we get back the descriptive error message:
<details>
  <summary>post-fix error message</summary>

```
Traceback (most recent call last):                 
  File "/home/xmfan/core/pytorch/torch/fx/graph_module.py", line 307, in __call__
    return super(self.cls, obj).__call__(*args, **kwargs)  # type: ignore[misc]
  File "/home/xmfan/core/pytorch/torch/nn/modules/module.py", line 1527, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)                                                                                                  
  File "/home/xmfan/core/pytorch/torch/nn/modules/module.py", line 1537, in _call_impl
    return forward_call(*args, **kwargs)                                                                                                     
  File "<eval_with_key>.0", line 4, in forward
    def forward(self, inputs, sizes, hooks):                                                                                                 
IndexError: list index out of range
                                                                                                                                             
Call using an FX-traced Module, line 4 of the traced Module's generated forward function:
                                                                                                                                             
def forward(self, inputs, sizes, hooks):
                                                                                                                                             
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
    getitem = inputs[0]                                                                                                                      
                                                                      
    getitem_1 = inputs[1];  inputs = None      
```

</details>

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

xmfan commented:
@pytorchbot merge
